### PR TITLE
Added Callbacks Fixes #5

### LINKED ISSUE #5: Add callbacks
There has to be a way to associate responses with a given request.

For example:
```
Twilio.recieveMessages("610-761-0083", function(messsage) {
  var body = message.body;
  var gif = Giphy.getFirst(body);

  var sender_number = message.from

  Twilio.sendMms(senderNumber, imageURL);
});
```

### PR DESCRIPTION
**Pull request by TheThirdOne**

### CODE REVIEW

File: baton/handler.go [hunk 1]
```diff
@@ -14,6 +15,8 @@ func Handler() *mux.Router {
 	return m
 }
 
+var i = 0
```
Comment (zachlatta) [inline baton/handler.go:18]: Is this temporary for now? Seems like this'd be the kind of thing we'd want a database for.

File: baton/handler.go [hunk 2]
```diff
@@ -27,5 +30,7 @@ func serveWs(w http.ResponseWriter, r *http.Request) {
 	c := conn{send: make(chan []byte, 256), ws: ws}
 	h.register <- c
 	go c.writePump()
+	c.send <- []byte(strconv.Itoa(i))
```
Comment (zachlatta) [inline baton/handler.go:33]: Might it make sense to send this in an initialization JSON object? Something like:

``` js
{
  "type": "initialization",
  "data": {
    "id": [id]
  }
}
```

Not saying that should be the format of the object, but just wondering if there should be an object sent at all instead of a raw number.

File: baton/modules/giphy/giphy.go [hunk 3]
```diff
@@ -55,16 +56,16 @@ func (g Giphy) RunCommand(cmd string, body interface{}, resp chan<- interface{})
...
+
+	rel, err := url.Parse(path)
+	if err != nil {
+		return url.URL{}, err
+	}
+
+	u := baseURL.ResolveReference(rel)
+	u.RawQuery = v.Encode()
+	return *u, nil
+}
```
Comment (zachlatta) [inline baton/modules/giphy/giphy.go:71]: In the future can we always make sure to `gofmt` files before committing? Not doing so makes diffs like these more difficult to read if changes were made.

**Comment by zachlatta** [comment]
Gonna go ahead and merge despite comments, I'm sure we can fix the concerns at a later time.

File: baton/handler.go (hunk 1 above)
Comment (TheThirdOne) [inline baton/handler.go:18]: This is simply something to give each user a unique identifier. No need to store put it in a database.

File: baton/handler.go (hunk 2 above)
Comment (TheThirdOne) [inline baton/handler.go:33]: If we end up doing more complicated initialization (we almost certainly will once the concept of users appears), this will become a json object, but for now, 1 expected value doesn't merit json.

File: baton/handler.go (hunk 1 above)
Comment (zachlatta) [inline baton/handler.go:18]: What happens when we fire up multiple instances of Maestro? Will there be collisions?

File: baton/handler.go (hunk 2 above)
Comment (zachlatta) [inline baton/handler.go:33]: Okay, before we deploy this I think I'd be most comfortable if all data sent over the wire was JSON.

Comment (TheThirdOne) [inline baton/handler.go:33]: ok

File: baton/handler.go (hunk 1 above)
Comment (TheThirdOne) [inline baton/handler.go:18]: If both instances are using the same port, yes. Otherwise, the client won't use both and it will be fine.

To be clear, this is an identifier per session not per user.

Comment (zachlatta) [inline baton/handler.go:18]: What if the two instances are behind a load balancer, like so?

```
 LB
 |
 /\
M  M
```

Comment (TheThirdOne) [inline baton/handler.go:18]: Several things are not set up to be ready for a load balancer. When/if we get to the point that a load balancer makes sense, I'll have to change it. 

I wouldn't expect we would need a load balancer any time in the immediate future. Also, load balancing websockets such that they would switch which machine they are messaging is probably not ideal. 

tldr: I'm not expecting us to use loadbalancing soon; if we do it so that websockets only connect to one machine at a time, we will be fine anyways.

Comment (zachlatta) [inline baton/handler.go:18]: I'd consider building a stateless app a code-quality thing, see the 12 factor application (http://12factor.net/).

Comment (zachlatta) [inline baton/handler.go:18]: See http://12factor.net/processes.
